Fix liquid ammo reload and container issues

akirashirosawa opened the pull request:
#### Summary
SUMMARY: Bugfixes "Fix liquid ammo reload and container issues"

#### Purpose of change
Fix bug: When you try to recharge a gasoline lantern or oil lamp, the liquid (gasoline or lamp oil) simply disappears, while the number of charges in the tools does not increase.
Fix? issue #40188 (can't completely reproduce).

#### Describe the solution
1. Added `debugmsg` with real reason when reload/recharge (`item::put_in`) failed. The real reason for failure pops up from the bottom up instead of the `"No Success"` stub. Now we can see error message when reloaded tool with liquid ammo:
<img width="998" alt="1" src="https://user-images.githubusercontent.com/38557723/86011560-7e13bb00-ba25-11ea-8bab-9173d37631c0.png">

2. Now we can see that the cause of the bug is PR #41519, because  ~~`LGTM, didn't test it (@KorGgenT)`~~ `pocket_data` of tools does not set `watertight` field when ammo is liquid. We could put the `watertight` and` airtight` fields everywhere, where liquid or gaseous ammo is present (hello PR #41640). But the problem is that the non-intuitive logic for processing the field `ammo_restriction`, and same errors were already and will occur again. So I changed the order if-else flow in the function `item_pocket::can_contain `, so that `ammo_restriction` overrides not only `volume`, `weight`, but also `watertight` and` airtight` to use the given ammo type. ~~As it doing `item_restriction` already (but this is not documented).~~

3. Update docs about `ammo_restriction` overrides `watertight`, `airtight`.

#### Describe alternatives you've considered
May add the `watertight` and` airtight` fields everywhere, where liquid or gaseous ammo (like in PR #41640). But I think this problem will up again and again, because intuitively it seems that `ammo_restriction` field is enough. And it’s not always clear what phase the ammo is in, for example in a lighter ammo is `butane `. Butane is gas, right, but is it liquid in the lighter?

#### Testing
Spawn some gasoline tool (like gasoline lantern) and gasoline for it, then try reload tool (successful after this fix).

#### Additional context
Adding the `debugmsg` in `item::put_in` showed a lot of bugs related to the container/pocket handling code. The errors and fixes are described in detail in these comments:

https://github.com/CleverRaven/Cataclysm-DDA/pull/41677#issuecomment-669063229:
1. item too big (fix #41928) 
2. item too heavy
3. is not a container

https://github.com/CleverRaven/Cataclysm-DDA/pull/41677#issuecomment-670167897:
3. is not a container (continuation)
4. holster already contains an item (item::use_charges)
5. tried to put an item (null)

https://github.com/CleverRaven/Cataclysm-DDA/pull/41677#issuecomment-670489885
6. holster already contains an item (player::gun_value)

https://github.com/CleverRaven/Cataclysm-DDA/pull/41677#issuecomment-670510332 
JSON error examples (for future fix, not in this PR) see issue #43149

anothersimulacrum commented on src/item_contents.cpp:
```diff
@@ -322,9 +324,11 @@ ret_val<item_pocket *> item_contents::find_pocket_for( const item &it,
         }
     }
-    return ret;
+    return ret_val<item_pocket *>::make_failure( null_pocket, _( msg ) );
```
Translation doesn't work like this, the string being translated needs to be available before compile time, which it is not here.

akirashirosawa replied:
I forgot about how it works, thanks, fixed it.

anothersimulacrum commented on src/item_contents.cpp:
```diff
@@ -307,10 +307,12 @@ ret_val<item_pocket *> item_contents::find_pocket_for( const item &it,
     static item_pocket *null_pocket = nullptr;
-    ret_val<item_pocket *> ret = ret_val<item_pocket *>::make_failure( null_pocket,
-                                 _( "is not a container" ) );
+    std::string msg = "is not a container";
```
This needs to 
```suggestion
    std::string msg = _( "is not a container" );
```
to be translated.
You can figure something else out for the comparison you do below.

akirashirosawa commented:
I see that the tests failed. Everything works fine on my macOS. I’ll try to find out what the problem is.

akirashirosawa commented:
Tests work as before, they just now display an error, where previously they did not output anything. I just switch to master and added
```cpp
void item::put_in( const item &payload, item_pocket::pocket_type pk_type )
{
    auto ret = contents.insert_item( payload, pk_type );
    if( !ret.success() ) {
        debugmsg( "put_in fail");
    }
    on_contents_changed();
}
```
And got  flood "put_in fail" when run `make check`. But everything works well in the game. Maybe something is wrong with the tests? I can simply commented `debugmsg` in the `put_in` function. Anyway, your vote in a plebiscite means nothing and put_in does not fail so simply (Russian joke).

ZhilkinSerg commented:
Needs conflict resolution.

akirashirosawa commented:
@ZhilkinSerg done.

In this merge I have preferred
`return ret_val<bool>::make_failure( pocket.str() );`
instead of
`return ret_val<bool>::make_failure( "Failed to insert into pocket" );`
because messages like:
`tried to put an item (vitamins) in a container (bottle_plastic_small) that cannot contain it: pocket unacceptable because: item too big`
`tried to put an item (chem_zinc_powder) in a container (bottle_plastic_small) that cannot contain it: pocket unacceptable because: item is too heavy`
`tried to put an item (software_useless) in a container (usb_drive) that cannot contain it: is not a container`
shows real reason and more useful than:
`(...) that cannot contain it: Failed to insert into pocket`.

Unfortunately, all the messages and reasons are real (there are problems with chemistry and drugs containers, see issue #41928, and problem with `usb_drive`, that I did not notice before).

ZhilkinSerg commented:
```
(shoulder_strap) in (remington_870): is not a container
(shoulder_strap) in (shotgun_d): is not a container
(shoulder_strap) in (mossberg_500): is not a container
(shoulder_strap) in (shotgun_410): is not a container
(shoulder_strap) in (mossberg_500_security): is not a container
(shoulder_strap) in (browning_a5): is not a container
(shoulder_strap) in (mossberg_930): is not a container
(shoulder_strap) in (shotgun_s): is not a container
(shoulder_strap) in (remington_1100): is not a container
(shoulder_strap) in (mossberg_500_security): is not a container4.278 s: grenade_lethality
(light_plus_battery_cell) in (cell_phone): is not a container
(shoulder_strap) in (crossbow): is not a container
(medium_battery_cell) in (chemistry_set): is not a container
(medium_battery_cell) in (welder): is not a container
(medium_disposable_cell) in (foodperson_mask): is not a container
(light_disposable_cell) in (two_way_radio): is not a container
(shoulder_strap) in (m4a1): is not a container
(holo_sight) in (m4a1): is not a container
(shoulder_strap) in (hk_mp5): is not a container
(suppressor) in (hk_mp5): is not a container
(holo_sight) in (hk_mp5): is not a container
(light_disposable_cell) in (flashlight): is not a container
(rail_laser_sight) in (hk_mp5): is not a container
(shoulder_strap) in (ksg): is not a container
(M24) in (back_holster): pocket unacceptable because item is too long
(medium_disposable_cell) in (tazer): is not a container
(shoulder_strap) in (remington700_270): is not a container
(light_plus_battery_cell) in (portable_game): is not a container
(shoulder_strap) in (h&k416a5): is not a container
(acog_scope) in (h&k416a5): is not a container
(medium_plus_battery_cell) in (mil_mess_kit): is not a container
(shoulder_strap) in (m2010): is not a container
(rifle_scope) in (m2010): is not a container
(bipod) in (m2010): is not a container
(light_plus_battery_cell) in (mp3): is not a container
(shoulder_strap) in (rifle_flintlock): is not a container
(light_disposable_cell) in (camera_pro): is not a container
(medium_battery_cell) in (mess_kit): is not a container
(light_minus_disposable_cell) in (miner_hat): is not a container
```

akirashirosawa commented:
@ZhilkinSerg I didn't add all these errors, like `item too big`, `item is too heavy`, `is not a container`, I just **let them pop up** in this PR. I was going to create separate issues/fixes PR for each type of errors, but if you insist, okay, I'll do everything here.

### 1. `item too big`
`(gummy_vitamins) in (bottle_plastic_small): item too big`
`(weak_antibiotic) in (bottle_plastic_small): item too big`
`(calcium_tablet) in (bottle_plastic_small): item too big`
`(vitamins) in (bottle_plastic_small): item too big`

I assumed (here issue #41928) that the reason was that the meds and chems was not made of soft material. The reason was not in JSON, but in the code. Instead of checking one tablet fits into the bottleneck, it is checked 20 tablets fit into the bottleneck at once.
Fix:
```diff
--- a/src/item_pocket.cpp
+++ b/src/item_pocket.cpp
@@ -1007,10 +1007,11 @@ ret_val<item_pocket::contain_code> item_pocket::can_contain( const item &it ) co
 
     // liquids and gases avoid the size limit altogether
     // soft items also avoid the size limit
+    // count_by_charges items check volume of single charge
     if( !it.made_of( phase_id::LIQUID ) && !it.made_of( phase_id::GAS ) &&
         !it.is_frozen_liquid() &&
         !it.is_soft() && data->max_item_volume &&
-        it.volume() > *data->max_item_volume ) {
+        !it.charges_per_volume( *data->max_item_volume ) ) {
         return ret_val<item_pocket::contain_code>::make_failure(
                    contain_code::ERR_TOO_BIG, _( "item too big" ) );
     }

```

### 2. `item too heavy`
`(chem_chromium_oxide) in (bottle_plastic_small): item is too heavy`
`(chem_manganese_dioxide) in (bottle_plastic_small): item is too heavy`
`(chem_zinc_powder) in (bottle_plastic_small): item is too heavy`
Add more info to messages:
```diff
--- a/src/item.cpp
+++ b/src/item.cpp
@@ -1091,8 +1091,8 @@ ret_val<bool> item::put_in( const item &payload, item_pocket::pocket_type pk_typ
 {
     ret_val<bool> result = contents.insert_item( payload, pk_type );
     if( !result.success() ) {
-        debugmsg( "tried to put an item (%s) in a container (%s) that cannot contain it: %s",
-                  payload.typeId().str(), typeId().str(), result.str() );
+        debugmsg( "tried to put an item (%s) count (%d) in a container (%s) that cannot contain it: %s",
+                  payload.typeId().str(), payload.count(), typeId().str(), result.str() );
     }
     if( pk_type == item_pocket::pocket_type::MOD ) {
         update_modified_pockets();
@@ -8276,6 +8276,11 @@ units::volume item::get_total_capacity() const
     return contents.total_container_capacity();
 }
```
Now we see what the problem is with the number of charges:
`(chem_chromium_oxide) (709) in (bottle_plastic_small): item is too heavy`
`(chem_manganese_dioxide) (1462) in (bottle_plastic_small): item is too heavy`
`(chem_zinc_powder) (2033) in (bottle_plastic_small): item is too heavy`

The maximum charges are calculated based on the volume only. Need add a weight check.

Fix:
```diff
--- a/src/item.h
+++ b/src/item.h
@@ -735,6 +735,7 @@ class item : public visitable<item>
          * ammo, magazines, weapons, etc.
          */
         units::volume get_total_capacity() const;
+        units::mass get_total_weight_capacity() const;
 
         // recusive function that checks pockets for remaining free space
         units::volume check_for_free_space( const item *it ) const;
--- a/src/item.cpp
+++ b/src/item.cpp
@@ -8276,6 +8276,11 @@ units::volume item::get_total_capacity() const
     return contents.total_container_capacity();
 }
 
+units::mass item::get_total_weight_capacity() const
+{
+    return contents.total_container_weight_capacity();
+}
+
 int item::get_remaining_capacity_for_liquid( const item &liquid, bool allow_bucket,
         std::string *err ) const
 {
--- a/src/item_group.cpp
+++ b/src/item_group.cpp
@@ -309,7 +309,12 @@ void Item_modifier::modify( item &new_item ) const
 
     if( max_capacity == -1 && !cont.is_null() && ( new_item.made_of( phase_id::LIQUID ) ||
             ( !new_item.is_tool() && !new_item.is_gun() && !new_item.is_magazine() ) ) ) {
-        max_capacity = new_item.charges_per_volume( cont.get_total_capacity() );
+        if( new_item.type->weight == 0_gram ) {
+            max_capacity = new_item.charges_per_volume( cont.get_total_capacity() );
+        } else {
+            max_capacity = std::min( new_item.charges_per_volume( cont.get_total_capacity() ),
+                                     new_item.charges_per_weight( cont.get_total_weight_capacity() ) );
+        }
     }
 
     const bool charges_not_set = charges.first == -1 && charges.second == -1;
```

### 3. `is not a container`
`(software_medical) in (usb_drive): is not a container`
`(light_plus_battery_cell) in (cell_phone): is not a container`
`(medium_battery_cell) in (chemistry_set): is not a container`
`(shoulder_strap) in (ar15): is not a container`

Let's replace the message `is not a container` with something more useful:
```diff
--- a/src/item_contents.cpp
+++ b/src/item_contents.cpp
@@ -357,7 +357,7 @@ struct item_contents::item_contents_helper {
         }
 
         if( failure_messages.empty() ) {
-            return ret_val<my_pocket_type *>::make_failure( null_pocket, _( "is not a container" ) );
+            return ret_val<my_pocket_type *>::make_failure( null_pocket, _( "pocket with type (%d) not found"),  pk_type );
         }
         std::sort( failure_messages.begin(), failure_messages.end(), localized_compare );
         failure_messages.erase(

```
Now we can see that problem is related to the wrong `pocket_type`:
`(software_useless) in (usb_drive): pocket with type (0) not found`
`(shoulder_strap) in (ar15): pocket with type (0) not found`

0 == `item_pocket::pocket_type::CONTAINER` But in the first case (`software_useless`), was expected `SOFTWARE`( == 5), in second case (`shoulder_strap`) - `MOD` (== 3). I think, game call `item::put_in` with wrong `pk_type`.

Fix: work in progress...

ZhilkinSerg commented:
> I was going to create separate issues/fixes PR for each type of errors, but if you insist, okay, I'll do everything here.

Nah, that's okay top fix them in follow-up PRs. This list is just for future reference.

ZhilkinSerg commented:
Please use `io::enum_to_string<item_pocket::pocket_type>( pk_type ) );` for error messages, so we could see e.g. `CONTAINER` string instead of its internal number representation.

akirashirosawa commented:
> Please use `io::enum_to_string<item_pocket::pocket_type>( pk_type ) );` for error messages, so we could see e.g. `CONTAINER` string instead of its internal number representation.

Thank you. I googled how to output the names of the enums and without finding a good solution, I almost wrote own switch-case for this :)

akirashirosawa commented:
Now get:
`tried to put an item (software_useless) count (1) in a container (usb_drive) that cannot contain it: pocket with type (CONTAINER) not found`

And I found problem:
`src/item_group.cpp:420 [void Item_modifier::modify(item &) const] usb_drive`

```diff
--- a/src/item_group.cpp
+++ b/src/item_group.cpp
@@ -408,6 +413,9 @@ void Item_modifier::modify( item &new_item ) const
     }
 
     if( !cont.is_null() ) {
+        if( new_item.type->default_container->str() == "usb_drive" ) {
+            debugmsg( "usb_drive" );
+        }
         cont.put_in( new_item, item_pocket::pocket_type::CONTAINER );
         new_item = cont;
         if( sealed ) {
```

Here `put_in` function is called with `pocket_type::CONTAINER` for all items. Apparently this is causes errors not only for `software` in `usb_drive`, but also for `light_plus_battery_cell`  in `cell_phone` (expected `MAGAZINE_WELL`) and for `shoulder_strap` in `ar15` (expected `MOD`).

I haven't figured out how to fix it yet. Apparently, need to get `pocket_type` from `item`. But if it's here, `put_in` function would not need the second argument at all.

akirashirosawa commented:
### 3. `is not a container` (continuation)
I overloaded `put_in` to call it without wrong `item_pocket::pocket_type::CONTAINER` when spawning items.
Fix:
```diff
--- a/src/item_group.cpp
+++ b/src/item_group.cpp
@@ -148,7 +148,7 @@ Item_spawn_data::ItemList Single_item_creator::create( const time_point &birthda
     if( container_item ) {
         item ctr( *container_item, birthday );
         for( const item &it : result ) {
-            ctr.put_in( it, item_pocket::pocket_type::CONTAINER );
+            ctr.put_in( it );
         }
         result = ItemList{ ctr };
     }
@@ -408,7 +413,7 @@ void Item_modifier::modify( item &new_item ) const
     }
 
     if( !cont.is_null() ) {
-        cont.put_in( new_item, item_pocket::pocket_type::CONTAINER );
+        cont.put_in( new_item );
         new_item = cont;
         if( sealed ) {
             new_item.seal();
@@ -418,7 +423,7 @@ void Item_modifier::modify( item &new_item ) const
     if( contents != nullptr ) {
         Item_spawn_data::ItemList contentitems = contents->create( new_item.birthday() );
         for( const item &it : contentitems ) {
-            new_item.put_in( it, item_pocket::pocket_type::CONTAINER );
+            new_item.put_in( it );
         }
         if( sealed ) {
             new_item.seal();
@@ -637,7 +642,7 @@ item_group::ItemList item_group::items_from( const Group_tag &group_id, const ti
     if( group->container_item ) {
         item ctr( *group->container_item, birthday );
         for( const item &it : created ) {
-            ctr.put_in( it, item_pocket::pocket_type::CONTAINER );
+            ctr.put_in( it );
         }
         created = ItemList{ ctr };
     }

--- a/src/item.h
+++ b/src/item.h
@@ -324,6 +324,7 @@ class item : public visitable<item>
         bool ready_to_revive( const tripoint &pos ) const;
 
         bool is_money() const;
+        bool is_software() const;
 
@@ -747,6 +749,7 @@ class item : public visitable<item>
          * Puts the given item into this one.
          */
         ret_val<bool> put_in( const item &payload, item_pocket::pocket_type pk_type );
+        ret_val<bool> put_in( const item &payload );

--- a/src/item.cpp
+++ b/src/item.cpp
@@ -1101,6 +1101,21 @@ ret_val<bool> item::put_in( const item &payload, item_pocket::pocket_type pk_typ
     return result;
 }
 
+ret_val<bool> item::put_in( const item &payload )
+{
+    item_pocket::pocket_type pk_type = item_pocket::pocket_type::CONTAINER;
+    if( ( is_gun() && payload.is_gunmod() ) || ( is_tool() && payload.is_toolmod() ) ) {
+        pk_type = item_pocket::pocket_type::MOD;
+    } else if( typeId().str() == "usb_drive" && payload.is_software() ) {
+        pk_type = item_pocket::pocket_type::SOFTWARE;
+    } else if( ( is_gun() || is_tool() ) && ( payload.is_battery() || payload.is_magazine() ) ) {
+        pk_type = item_pocket::pocket_type::MAGAZINE_WELL;
+    } else if( ( is_magazine() || is_gun() ) && ( payload.is_ammo() ) ) {
+        pk_type = item_pocket::pocket_type::MAGAZINE;
+    }
+    return put_in( payload, pk_type );
+}
+
@@ -5967,6 +5982,11 @@ bool item::is_money() const
     return ammo_types().count( ammotype( "money" ) );
 }
 
+bool item::is_software() const
+{
+    return type->default_container->str() == "usb_drive";
+}

```

### 4. **NEW** `holster already contains an item`
`(medium_battery_cell) in (chemistry_set): holster already contains an item`
`(heavy_plus_battery_cell) in (UPS_off): holster already contains an item`

The problems that going visible now, because I added `debugmsg ` seems to be endless...(
This time, popping up error when character try to craft something, that consume tool charges. `use_charges` function tries to `put_in` a battery into the tool when it already contains it. Seems `item temp (* e);` copy/create item with magazine inside. Don't need to create magazine additionally and then try to insert it. Also may call `ammo_set` directly on the item, it will detect magazine by self and even create it if necessary.
Fix:
```diff
--- a/src/item.cpp
+++ b/src/item.cpp
@@ -8616,18 +8641,12 @@ bool item::use_charges( const itype_id &what, int &qty, std::list<item> &used,
                 int n = std::min( e->ammo_remaining(), qty );
                 qty -= n;
 
-                item temp( *e );
-                if( temp.is_magazine() ) {
+                if( e->is_magazine() || e->magazine_current() != nullptr ) {
+                    item temp( *e );
                     temp.ammo_set( e->ammo_current(), n );
-                } else if( e->magazine_current() != nullptr ) {
-                    item temp_mag( *e->magazine_current() );
-                    temp_mag.ammo_set( e->ammo_current(), n );
-                    temp.put_in( temp_mag, item_pocket::pocket_type::MAGAZINE_WELL ); // <-- throw error holster already contains an item
-                } else {
-                    return VisitResponse::SKIP;
+                    used.push_back( temp );
+                    e->ammo_consume( n, pos );
                 }
-                used.push_back( temp );
-                e->ammo_consume( n, pos );
             }
             return VisitResponse::SKIP;
```

akirashirosawa commented:
### 5. `tried to put an item (null)`

`tried to put an item (null) in (EMPbomb): pocket with type (MAGAZINE) not found`
`tried to put an item (null) in (grenade): pocket with type (MAGAZINE) not found`
`tried to put an item (null) in (dynamite): pocket with type (MAGAZINE) not found`

Function `mx_supplydrop` spawn `item_group` `"grenades"` by call func `place_items` with option `magazine == 100`. 
```cpp
place_items( item_group, 80, *p, *p, true, calendar::start_of_cataclysm, 100 )
```
Then `place_items` tries to `put_in` default magazine into the grenade and fail.

Added check for `magazine_default` and `ammo_default` exists.
Fix:
```diff
--- a/src/mapgen.cpp
+++ b/src/mapgen.cpp
@@ -5756,10 +5756,11 @@ std::vector<item *> map::place_items( const items_location &loc, const int chanc
     }
     for( item *e : res ) {
         if( e->is_tool() || e->is_gun() || e->is_magazine() ) {
-            if( rng( 0, 99 ) < magazine && !e->magazine_integral() && !e->magazine_current() ) {
+            if( rng( 0, 99 ) < magazine && e->magazine_default() && !e->magazine_integral() &&
+                !e->magazine_current() ) {
                 e->put_in( item( e->magazine_default(), e->birthday() ), item_pocket::pocket_type::MAGAZINE );
             }
-            if( rng( 0, 99 ) < ammo && e->ammo_remaining() == 0 ) {
+            if( rng( 0, 99 ) < ammo && e->ammo_default() && e->ammo_remaining() == 0 ) {
                 e->ammo_set( e->ammo_default() );
             }
         }
```

akirashirosawa commented:
### 6. `holster already contains an item` (`player::gun_value`)

`(wp22mag) in (walther_p22): holster already contains an item`
`(stanag30) in (ar_pistol): holster already contains an item`
`(uzimag) in (uzi): holster already contains an item`

Like function `item::use_charges`, `player::gun_value` create item with magazine inside, new magazine and then try to insert it. Same fix as above: `ammo_set` use current magazine or create default magazine if needed.
Fix:
```diff
--- a/src/ranged.cpp
+++ b/src/ranged.cpp
@@ -1766,17 +1766,8 @@ double player::gun_value( const item &weap, int ammo ) const
 
     damage_instance gun_damage = weap.gun_damage();
     item tmp = weap;
-    if( tmp.is_magazine() ) {
+    if( tmp.is_magazine() || tmp.magazine_current() || tmp.magazine_default() ) {
         tmp.ammo_set( ammo_type );
-    } else if( tmp.contents.has_pocket_type( item_pocket::pocket_type::MAGAZINE_WELL ) ) {
-        item mag;
-        if( weap.magazine_current() ) {
-            mag = item( *weap.magazine_current() );
-        } else {
-            mag = item( weap.magazine_default() );
-        }
-        mag.ammo_set( ammo_type );
-        tmp.put_in( mag, item_pocket::pocket_type::MAGAZINE_WELL ); // <-- throw error holster already contains an item
     }
     int total_dispersion = get_weapon_dispersion( tmp ).max() +
                            effective_dispersion( tmp.sight_dispersion() );
```

akirashirosawa commented:
Errors in the code, that I managed to find with tests seems ended. There are many more errors in JSON: items are too long, big or heavy for their containers, some professions do not have enough space in their inventory for items, etc.
JSON error examples:
`(M24) in (back_holster): item is too long`
```json
"id": "sniper_police",
{ "item": "M24", "ammo-item": "762_51", "charges": 20, "container-item": "back_holster" }
--------------------------
"id": "back_holster",
"max_item_length": "60 cm",
------------------------
"id": "M24",
"longest_side": "992 mm", // 99.2 cm > 60 cm
```

`(fertilizer_commercial) count (60) in (bag_canvas): item is too heavy`

```json
"id": "fertilizer_commercial",
"weight": "350 g",
"container": "bag_canvas",
"volume": "2 L",
"charges": 60,
------------------------
"id": "bag_canvas",
"max_contains_weight": "15 kg" // 0,350 * 60 = 21 kg > 15 kg
```

`(crackers) count (8) in (box_small): item too big`

```json
"id": "crackers",
"weight": "56 g",
"container": "box_small",
"volume": "1 L",
"charges": 8,
------------------------
"id": "box_small",
"max_contains_volume": "990 ml" //  1 L > 990 ml
// Many items with default container "box_small" have a volume == 1 L.
// And volume reduction to 990 ml breaks a lot of things.
```

 I will add all the code fixes described in above comments to this PR and start work with JSON. I want to write script to check JSON items, containers, item_groups and will try to fix em all in a next PR(s).

akirashirosawa commented:
@ZhilkinSerg Ready to merge, i think. Tests will fail because of JSON (`M24` too long for `back_holster`, `box_small` too small, see comment above).

ZhilkinSerg commented on src/item.cpp:
```diff
@@ -1090,13 +1090,32 @@ bool item::merge_charges( const item &rhs )
     return result;
 }
 
+ret_val<bool> item::put_in( const item &payload )
```
There is a very similar , but not identical `item::migrate_content_item` function which should have the same logic as `item::put_in` - we should update it probably to get rid of duplicate code (not necessary in current PR).

KorGgenT replied:
i don't think this function should exist at all, because picking an implicit pocket is very dangerous.

akirashirosawa replied:
> i don't think this function should exist at all, because picking an implicit pocket is very dangerous.

@KorGgenT  `item::migrate_content_item` or `item::put_in`?
`item::put_in` w/o `pk_type` is safe as long as there are no rifles with a backpack that stores spare mods or magazines. For spawning such items (with many pocket types for one item type), really need to change itemgroup to specify a specific pocket or pocket type. But this PR is just a fix. It makes the game work, not improves it. Let improving itemgroup in another PR.

kevingranade replied:
If this is a bandaid bugfix it shouldn't be in the item interface, make it a helper function local to item_group that guesses the pocket type and only use it there, that leaves less code to rip back out later and doesn't risk other people using it.

kevingranade replied:
To clarify, invoking it should look like
```
const item_pocket::pocket_type pk_type = guess_pocket_for( container, new_item );
container.put_in( new_item, pk_type );
```

akirashirosawa replied:
Okay, I agree.

akirashirosawa replied:
Done.

ZhilkinSerg commented on src/item.cpp:
```diff
@@ -1090,13 +1090,32 @@ bool item::merge_charges( const item &rhs )
+    item_pocket::pocket_type pk_type = item_pocket::pocket_type::CONTAINER;
+    if( ( is_gun() && payload.is_gunmod() ) || ( is_tool() && payload.is_toolmod() ) ) {
+        pk_type = item_pocket::pocket_type::MOD;
+    } else if( typeId().str() == "usb_drive" && payload.is_software() ) {
```
We better extract `typeId().str() == "usb_drive"` check (here and in all other places) to a dedicated function (e.g. `is_software_storage`) (not necessary in current PR).

KorGgenT replied:
yeah this is a hack that completely bypasses the SOFTWARE pocket type.

akirashirosawa replied:
```cpp
bool item::is_software_storage() const
{
    const itype_id itype_usb_drive( "usb_drive" );
    return typeId() == itype_usb_drive;
}
```
Is that ok?

akirashirosawa replied:
```suggestion
    } else if( is_software_storage() && payload.is_software() ) {
```

akirashirosawa replied:
Updated the code.

ZhilkinSerg commented on src/item.cpp:
```diff
@@ -1090,13 +1090,32 @@ bool item::merge_charges( const item &rhs )
+        pk_type = item_pocket::pocket_type::SOFTWARE;
+    } else if( ( is_gun() || is_tool() ) && ( payload.is_battery() || payload.is_magazine() ) ) {
+        pk_type = item_pocket::pocket_type::MAGAZINE_WELL;
+    } else if( ( is_magazine() || is_gun() ) && ( payload.is_ammo() ) ) {
```
```suggestion
    } else if( ( is_magazine() || is_gun() ) && payload.is_ammo() ) {
```

KorGgenT commented on src/item.cpp:
```diff
@@ -5963,6 +5982,11 @@ bool item::is_money() const
 
+bool item::is_software() const
+{
+    return type->default_container->str() == "usb_drive";
```
i disapprove of this to check against itype_id.

akirashirosawa replied:
```suggestion
    cata::optional<itype_id> optional_itype_usb_drive( "usb_drive" );
    return type->default_container == optional_itype_usb_drive;
```
That's better?

akirashirosawa replied:
Updated the code.

KorGgenT commented on src/item_group.cpp:
```diff
@@ -148,7 +148,7 @@ Item_spawn_data::ItemList Single_item_creator::create( const time_point &birthda
         item ctr( *container_item, birthday );
         for( const item &it : result ) {
-            ctr.put_in( it, item_pocket::pocket_type::CONTAINER );
+            ctr.put_in( it );
```
rather than this maybe we should have the itemgroup allow you to specify the pocket type.

akirashirosawa replied:
It's probably better. This will allow to spawn items with different pocket types for the same item type. But let it be a separate PR. This PR is only fix for make game playable.

akirashirosawa replied:
Removed put_in and added guess_pocket_for

KorGgenT commented on src/ranged.cpp:
```diff
@@ -1766,17 +1766,8 @@ double player::gun_value( const item &weap, int ammo ) const
     damage_instance gun_damage = weap.gun_damage();
     item tmp = weap;
-    if( tmp.is_magazine() ) {
+    if( tmp.is_magazine() || tmp.magazine_current() || tmp.magazine_default() ) {
```
you sure about this? ammo_set() does not currently handle setting a magazine item and then setting ammo inside an item.

akirashirosawa replied:
I double checked `ammo_set`. As the comment says:
```cpp
/**
 * Filter setting the ammo for this instance
 * Any existing ammo is removed. If necessary a magazine is also added.
 * @param ammo specific type of ammo (must be compatible with item ammo type)
 * @param qty maximum ammo (capped by item capacity) or negative to fill to capacity
 * @return same instance to allow method chaining
 */
item &ammo_set( const itype_id &ammo, int qty = -1 );
```
This function really if necessary creates a magazine and inserts it into an item, then fills it with ammo.
```cpp
// ...
if( !magazine_current() ) {
	const itype *mag = find_type( magazine_default() );
	// ...
	put_in( item( mag ), item_pocket::pocket_type::MAGAZINE_WELL );
}
item *mag_cur = magazine_current();
if( mag_cur != nullptr ) {
    mag_cur->ammo_set( ammo, qty );
}
```

akirashirosawa commented:
> I feel like we're backpedaling in a couple areas, and those areas should get another iteration to improve them. You're right that moving airtight/watertight to be overridden by ammo_restriction is the right move, however.

@KorGgenT These PR just bug fixes. I do not have the resources to rewrite itemgroup handling (I don't know the game code enough). For now is not danger of defining `pocket_type` based on the type of container and item. This fixes problems with spawn items. This does not create new problems. This is totally better than it was. I think this is enough for a bug fix PR. The next step, in the next PRs, I think, fixes JSON problems listed here (https://github.com/CleverRaven/Cataclysm-DDA/pull/41677#issuecomment-670510332).

akirashirosawa commented on src/item.h:
```diff
@@ -324,6 +324,7 @@ class item : public visitable<item>
         bool ready_to_revive( const tripoint &pos ) const;
 
         bool is_money() const;
+        bool is_software() const;
```
```suggestion
        bool is_software() const;
        bool is_software_storage() const;
```
Add `is_software_storage`.

akirashirosawa commented on src/item.cpp:
```diff
@@ -5963,6 +5982,11 @@ bool item::is_money() const
 
+bool item::is_software() const
+{
+    return type->default_container->str() == "usb_drive";
```
```suggestion
    cata::optional<itype_id> optional_itype_usb_drive( "usb_drive" );
    return type->default_container == optional_itype_usb_drive;
}

bool item::is_software_storage() const
{
    const itype_id itype_usb_drive( "usb_drive" );
    return typeId() == itype_usb_drive;
```
Change `itype_id` check. Add `item::is_software_storage()`.

akirashirosawa commented on src/item.cpp:
```diff
@@ -1090,13 +1090,32 @@ bool item::merge_charges( const item &rhs )
+        pk_type = item_pocket::pocket_type::MOD;
+    } else if( typeId().str() == "usb_drive" && payload.is_software() ) {
+        pk_type = item_pocket::pocket_type::SOFTWARE;
+    } else if( ( is_gun() || is_tool() ) && ( payload.is_battery() || payload.is_magazine() ) ) {
```
```suggestion
    } else if( ( is_gun() || is_tool() ) && payload.is_magazine() ) {
```
battery is magazine too.

akirashirosawa commented on src/item.cpp:
```diff
@@ -1090,13 +1090,32 @@ bool item::merge_charges( const item &rhs )
+        pk_type = item_pocket::pocket_type::SOFTWARE;
+    } else if( ( is_gun() || is_tool() ) && ( payload.is_battery() || payload.is_magazine() ) ) {
+        pk_type = item_pocket::pocket_type::MAGAZINE_WELL;
+    } else if( ( is_magazine() || is_gun() ) && payload.is_ammo() ) {
```
```suggestion
    } else if( ( is_magazine() ) && payload.is_ammo() ) {
```
gun is feeds by ammo is magazine too

esotericist commented:
i don't have a specific opinion on the code in this (since i don't deal with this aspect of the game if i can manage to avoid it), but i think there's a critical viewpoint problem that i'm seeing repeated in posts from @akirashirosawa 

> These PR just bug fixes. I do not have the resources to rewrite itemgroup handling (I don't know the game code enough). For now is not danger of defining pocket_type based on the type of container and item. This fixes problems with spawn items. This does not create new problems. This is totally better than it was. I think this is enough for a bug fix PR.

due to the nature of the game, maintainability is crucial. a "bugfix" that creates substantial work that needs to be done later, for a non-critical bug, *is* creating new problems.

if you fix a bug in the wrong way, at the wrong scope, and do so in a fashion that will make it more difficult to fix it correctly later? That isn't fixing the game, that's breaking it harder, and thus by definition is not better than it was.

akirashirosawa commented:
@esotericist 
> will make it more difficult to fix it correctly later?

This fix doesn't make future fixes more difficult.
>"bugfix" that creates substantial work that needs to be done later, for a non-critical bug, is creating new problems.

This fix doesn't creates substantial work that needs to be done later. There are two ways to fix this issues. Simple and complex. It's simple. But it doesn't override or make the complex one harder. I think, when in the game can't spawn big list of items (tools with batteries, guns with mods and magazine, magazines with ammo, software) kind a critical.

But I agree with @kevingranade, that it was not worth adding a new function to the `item` interface. I moved it to the `item_group` local.

KorGgenT commented on src/item_group.cpp:
```diff
@@ -103,6 +103,23 @@ item Single_item_creator::create_single( const time_point &birthday, RecursionLi
     return tmp;
 }
 
+item_pocket::pocket_type guess_pocket_for( const item &container, const item &payload )
```
```suggestion
static pocket_type guess_pocket_for( const item &container, const item &payload )
```

akirashirosawa commented on src/item_group.cpp:
```diff
@@ -103,6 +103,23 @@ item Single_item_creator::create_single( const time_point &birthday, RecursionLi
     return tmp;
 }
 
+static pocket_type guess_pocket_for( const item &container, const item &payload )
```
@KorGgenT  It throws error. You mean:
```suggestion
static item_pocket::pocket_type guess_pocket_for( const item &container, const item &payload )
```
? I'm not a C++ programmer, sorry.

KorGgenT replied:
no i do not.
I don't know what error you're seeing, but a `static` function is a function that is only available in that one cpp file. what Kevin said about scope is correct; we want to keep it as small as possible. defining a static function like that also means it is not a part of the class, so you can't really use class member unless you're passing itself as an argument. making it a static function is what i thought we were gonna do when kevin mentioned scoping it basically only to item groups. it'll also discourage people from reusing the function while we're still waiting on a more complete resolution which i mentioned on an earlier comment about creating a new json member in item groups for pocket type.

akirashirosawa replied:
src/item_group.cpp:106:8: error: unknown type name 'pocket_type'; did you mean 'item_pocket::pocket_type'?
static pocket_type guess_pocket_for( const item &container, const item &payload )
       ^~~~~~~~~~~
       item_pocket::pocket_type
src/item_pocket.h:33:14: note: 'item_pocket::pocket_type' declared here
        enum pocket_type {
             ^
1 error generated.
make: *** [obj/tiles/item_group.o] Error 1

KorGgenT replied:
argh my mistake. i apologize. i really need to wait to review things until after i've woken up fully. anyway, the static makes it good now, clang tidy won't yell you for it.

jbytheway commented on src/item_contents.cpp:
```diff
@@ -357,7 +357,8 @@ struct item_contents::item_contents_helper {
         if( failure_messages.empty() ) {
-            return ret_val<my_pocket_type *>::make_failure( null_pocket, _( "is not a container" ) );
+            return ret_val<my_pocket_type *>::make_failure( null_pocket, _( "pocket with type (%s) not found" ),
+                    io::enum_to_string<item_pocket::pocket_type>( pk_type ) );
```
```suggestion
                    io::enum_to_string( pk_type ) );
```
This doesn't need an explicit template argument; it will be deduced.

jbytheway commented on src/item_pocket.cpp:
```diff
@@ -938,38 +938,12 @@ ret_val<item_pocket::contain_code> item_pocket::can_contain( const item &it ) co
-            return ret_val<item_pocket::contain_code>::make_failure(
-                       contain_code::ERR_LIQUID, _( "can't contain liquid" ) );
-        }
-        if( size() != 0 && !item( contents.front() ).combine( it ) ) {
-            return ret_val<item_pocket::contain_code>::make_failure(
-                       contain_code::ERR_LIQUID, _( "can't mix liquid with contained item" ) );
-        }
```
Is it safe to move this check after the restriction-based checks?  I think this will allow putting two different liquids in a container when both fit the restriction, which we probably don't want.

akirashirosawa replied:
It safe, because the ammo match checked. Magazine can't contain different ammo, so different liquids too.
```cpp
// the ammo must match what's inside
        if( !contents.empty() ) {
            if( it_ammo != contents.front().ammo_type() ) {
                return ret_val<item_pocket::contain_code>::make_failure(
                           contain_code::ERR_AMMO, _( "item is not the correct ammo type" ) );
            } else {
                internal_count = contents.front().count();
            }
        }
```

jbytheway replied:
That code only applies to things with an ammo_restriction; what about things with an item_restriction?  The OP says that now also overrides watertight, but I don't see code changes to that effect.

jbytheway replied:
As a side note: I don't think it's as important to make this prioritization change since #42022.

akirashirosawa replied:
> what about things with an item_restriction?

I have not changed anything about `item_restriction`.

>As it doing item_restriction already (but this is not documented).

I added

>(and watertight, airtight ignores too). Overrides only if holster is true.

I mean this code:
```cpp
    if( !data->item_id_restriction.empty() &&
        data->item_id_restriction.count( it.typeId() ) == 0 ) {
        return ret_val<item_pocket::contain_code>::make_failure(
                   contain_code::ERR_FLAG, _( "holster does not accept this item type" ) );
    }
    if( data->holster && !contents.empty() ) {
        item item_copy( contents.front() );
        if( item_copy.combine( it ) ) {
            return ret_val<item_pocket::contain_code>::make_success();
```
But now I see that in fact the check is skipped, only if the holster contains the same object, which obviously has already passed the checks. I will delete my note about `item_restriction` from doc.

akirashirosawa replied:
> I don't think it's as important to make this prioritization change since #42022.

I think this is more intuitive logic. volume\weight\length not checked when specifying ammo. There is no need to check these parameters too. This will save from future mistakes. It also requires less JSON writing.

jbytheway commented on src/item.cpp:
```diff
@@ -5963,6 +5967,18 @@ bool item::is_money() const
     return ammo_types().count( ammotype( "money" ) );
 }
 
+bool item::is_software() const
+{
+    cata::optional<itype_id> optional_itype_usb_drive( "usb_drive" );
+    return type->default_container == optional_itype_usb_drive;
+}
+
+bool item::is_software_storage() const
+{
+    const itype_id itype_usb_drive( "usb_drive" );
+    return typeId() == itype_usb_drive;
+}
+
```
I think the best we can do for now is to have `is_software_storage` be calculated by checking "does this have a software pocket" and `is_software` be calculated by "does the default container satisfy `is_software_storage`".

Alternatively, invent a new flag and add it to the flags for the `software` item from which all the others inherit and use that to determine software-ness.

akirashirosawa replied:
Okay.
```suggestion
bool item::is_software() const
{
    const itype_id &cont_value = type->default_container.value_or( "null" );
    if( !cont_value.is_null() ) {
        return item( cont_value ).is_software_storage();
    }
    return false;
}

bool item::is_software_storage() const
{
    return contents.has_pocket_type( item_pocket::pocket_type::SOFTWARE );
}
```
It is right?

jbytheway replied:
Looks reasonable.

Minor nit: I'd write the first function like:
```c++
    if( const cata_optional<itype_id> &cont = type->default_container ) {
        return item( cont ).is_software_storage();
    }
    return false;
```

jbytheway commented on src/item.cpp:
```diff
@@ -5963,6 +5967,19 @@ bool item::is_money() const
 
+bool item::is_software() const
+{
+    const itype_id &cont_value = type->default_container.value_or( "null" );
+    if( !cont_value.is_null() ) {
+        return item( cont_value ).is_software_storage();
+    }
```
```suggestion
    if( const cata_optional<itype_id> &cont = type->default_container ) {
        return item( *cont ).is_software_storage();
    }
```

Just repeating this suggestion from my other comment which may now be hidden under an "Outdated" fold.

akirashirosawa replied:
```suggestion
    if( const cata::optional<itype_id> &cont = type->default_container ) {
        return item( *cont ).is_software_storage();
    }
```

akirashirosawa replied:
@jbytheway I rewrote your suggestion because of typo (`cata_optional` --> `cata::optional`). Actually, I don't really like such non-obvious code, but since you insist, let it be.

akirashirosawa commented:
@ZhilkinSerg I did rebase and squash all commits. 

After I fixed itemgroups (PR #43276), make it possible for them to create items with magazines and ammo, a new issue appeared in the tests. When it tried to insert magazine into gun with the magazine (`itemgroup` has already created it with a magazine). I fixed it.

I also commented the tests related to `item::put_in` that call `debugmsd`. I will uncomment them here when the PR #43363 will merged, or uncomment them in PR #43363 if this one is merged first.

All tests are passed, I think you can merge it. (only the test that does not apply to this PR fails `player_helpers.cpp:173: FAILED: REQUIRE( p.has_active_bionic( bioid ) )`).

ZhilkinSerg commented on doc/JSON_INFO.md:
```diff
@@ -2491,7 +2491,7 @@ Any Item can be a container. To add the ability to contain things to an item, yo
     "open_container": false,          // Default false. If true, the contents of this pocket will spill if this item is placed into another item.
     "fire_protection": false,         // Default false. If true, the pocket protects the contained items from exploding if tossed into a fire.
-    "ammo_restriction": { "ammotype": count }, // Restrict pocket to a given ammo type and count.  This overrides mandatory volume and weight to use the given ammo type instead.  A pocket can contain any number of unique ammotypes each with different counts, and the container will only hold one type (as of now).  If this is left out, it will be empty.
+    "ammo_restriction": { "ammotype": count }, // Restrict pocket to a given ammo type and count.  This overrides mandatory volume, weight and watertight, airtight to use the given ammo type instead.  A pocket can contain any number of unique ammotypes each with different counts, and the container will only hold one type (as of now).  If this is left out, it will be empty.
```
```suggestion
    "ammo_restriction": { "ammotype": count }, // Restrict pocket to a given ammo type and count.  This overrides mandatory volume, weight, watertight and airtight to use the given ammo type instead.  A pocket can contain any number of unique ammo types each with different counts, and the container will only hold one type (as of now).  If this is left out, it will be empty.
```

ZhilkinSerg commented:
Can you please rebase again and fix remaining TODOs from #43363 after it was merged?

akirashirosawa commented:
Done. I rebased branch and added debug message capture to tests.
